File: Java_or_Android/app/src/main/java/com/piyapan039285/java_or_android/Encryption.java
package com.piyapan039285.java_or_android;

import java.math.BigInteger;
import java.nio.charset.Charset;
import java.security.GeneralSecurityException;
import java.util.Random;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Created by Piyapan on 07/02/2017.
 */

public class Encryption
{
    public static String generateRandomHex(int byteLength)
    {
        int stringLength = byteLength * 2;

        String alphabet = "abcdef0123456789";
        String s = "";

        Random rand = new Random();

        for (int i = 0; i < stringLength; i++)
        {
            int r = rand.nextInt(alphabet.length());
            s += alphabet.charAt(r);
        }

        //prevent null block.
        s = s.replaceAll("00", "11");

        return s;
    }

    private static String dataToHexString(byte[] data) {
        StringBuilder str = new StringBuilder();
        for(int i = 0; i < data.length; i++)
            str.append(String.format("%02x", data[i]));

        String hexString = str.toString();
        return hexString;
    }

    private static byte[] dataFromHexString(String hexString) throws Exception
    {
        hexString = hexString.trim();
        hexString = hexString.replaceAll("[ ]", "");
        hexString = hexString.toLowerCase();

        String validHexChar = "abcdef0123456789";

        int len = hexString.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2)
        {
            char c1 = hexString.charAt(i);
            char c2 = hexString.charAt(i+1);

            if(!validHexChar.contains(""+c1) || !validHexChar.contains((""+c2)))
            {
                throw new Exception("Invalid encryption hex data");
            }

            data[i / 2] = (byte) ((Character.digit(c1, 16) << 4) + Character.digit(c2, 16));
        }
        return data;
    }

    public static String encryptData(String text, String hexKey) throws Exception
    {
        checkKey(hexKey);

        //generate random IV (16 bytes)
        String hexIV = generateRandomHex(16);

        //convert plainText to hex string.
        byte[] bytesData = text.getBytes(Charset.forName("UTF-8"));
        String hexStr = dataToHexString(bytesData);

        String cipherHexStr = __encryptData(hexStr, hexKey, hexIV);

        String hmacHexKey = generateRandomHex(16);
        String hmacHexStr = Encryption.__computeHMAC(hexIV, cipherHexStr, hexKey, hmacHexKey);

        String encryptedHexStr = hexIV + hmacHexKey + hmacHexStr + cipherHexStr;
        return encryptedHexStr;
    }

    public static String decryptData(String hexStr, String hexKey) throws Exception
    {
        checkKey(hexKey);

        String plainText = null;
        if (hexStr.length() > 128)
        {
            String hexIV = hexStr.substring(0, 32);
            String hmacHexKey = hexStr.substring(32, 64);
            String hmacHexStr = hexStr.substring(64, 128);
            String cipherHexStr = hexStr.substring(128);

            String computedHmacHexStr = Encryption.__computeHMAC(hexIV, cipherHexStr, hexKey, hmacHexKey);

            if (computedHmacHexStr.equalsIgnoreCase(hmacHexStr))
            {
                String decryptedStr = __decryptData(cipherHexStr, hexKey, hexIV);

                byte[] data = dataFromHexString(decryptedStr);
                plainText = new String(data, Charset.forName("UTF-8"));
            }
        }

        return plainText;
    }

    private static void checkKey(String hexKey) throws Exception
    {
        hexKey = hexKey.trim();
        hexKey = hexKey.replaceAll("[ ]", "");
        hexKey = hexKey.toLowerCase();

        if(hexKey.length() != 64)
        {
            throw new Exception("key length is not 256 bit (64 hex characters)");
        }

        int i;
        for(i=0;i<hexKey.length();i+=2)
        {
            if(hexKey.charAt(i) == '0' && hexKey.charAt(i+1) == '0')
            {
                throw new Exception("key cannot contain zero byte block");
            }
        }
    }

    private static String __computeHMAC(String hexIV, String cipherHexStr, String hexKey, String hmacHexKey) throws Exception
    {
        hexKey = hexKey.trim();
        hexKey = hexKey.replaceAll("[ ]", "");
        hexKey = hexKey.toLowerCase();

        hmacHexKey = hmacHexKey.toLowerCase();

        String hexString = hexIV + cipherHexStr + hexKey;
        hexString = hexString.toLowerCase();

        byte[] data = hexString.getBytes(Charset.forName("UTF-8"));
        byte[] hmacKey = hmacHexKey.getBytes(Charset.forName("UTF-8"));

        Mac sha256_HMAC = Mac.getInstance("HmacSHA256");
        SecretKeySpec secret_key = new SecretKeySpec(hmacKey, "HmacSHA256");
        sha256_HMAC.init(secret_key);

        byte[] hashbytes = sha256_HMAC.doFinal(data);
        String hashHexStr = Encryption.dataToHexString(hashbytes);

        return hashHexStr;
    }

    private static String __encryptData(String hexString, String hexKey, String hexIV) throws Exception
    {
        byte[] data = dataFromHexString(hexString);
        byte[] keyBytes = dataFromHexString(hexKey);
        byte[] iv = dataFromHexString(hexIV);

        SecretKeySpec key = new SecretKeySpec(keyBytes, "AES");

        //PKCS5Padding is PKCS #7 padding in java.
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv));

        byte[] encryptData = cipher.doFinal(data);

        String encryptHexData = dataToHexString(encryptData);
        return encryptHexData;
    }

    private static String __decryptData(String hexString, String hexKey, String hexIV) throws Exception
    {
        byte[] data = dataFromHexString(hexString);
        byte[] keyBytes = dataFromHexString(hexKey);
        byte[] iv = dataFromHexString(hexIV);

        SecretKeySpec key = new SecretKeySpec(keyBytes, "AES");

        //PKCS5Padding is PKCS #7 padding in java.
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv));

        byte[] decryptData = cipher.doFinal(data);

        String decryptHexData = dataToHexString(decryptData);
        return decryptHexData;
    }
}

File: Python/MultiPlatformAESEncryptionPython3.7/encryption.py
import math
import random
import codecs
from Crypto.Cipher import AES
from Crypto.Hash import HMAC, SHA256


def generate_random_hex(byte_length: int) -> str:
    string_length = byte_length * 2
    alphabet = "abcdef0123456789"
    s = ""

    for i in range(string_length):
        r = int(math.floor(random.random() * len(alphabet)))
        s += alphabet[r]

    # prevent null block.
    s = s.replace("00", "11")
    return s


def _data_from_hex_string(hex_string: str) -> bytes:
    hex_string = hex_string.strip().replace(' ', '').lower()
    alphabet = "abcdef0123456789"

    for ch in hex_string:
        if ch not in alphabet:
            raise Exception('Invalid encryption hex data')

    data_bytes = codecs.decode(hex_string, 'hex')
    return data_bytes


def data_to_hex_string(data_bytes: bytes) -> str:
    hex_string = codecs.encode(data_bytes, 'hex').decode()
    return hex_string


def encrypt_data(plain_text: str, hex_key: str) -> str:
    _check_key(hex_key)

    # generate random IV (16 bytes)
    hex_iv = generate_random_hex(16)

    # convert plainText to hex string.
    bytes_data = plain_text.encode('utf-8')
    hex_str = data_to_hex_string(bytes_data)

    cipher_hex_str = _encrypt_data(hex_str, hex_key, hex_iv)

    hmac_hex_key = generate_random_hex(16)
    hmac_hex_str = _compute_HMAC(hex_iv, cipher_hex_str, hex_key, hmac_hex_key)

    encrypted_hex_str = hex_iv + hmac_hex_key + hmac_hex_str + cipher_hex_str
    return encrypted_hex_str


def decrypt_data(hex_str: str, hex_key: str) -> str:
    _check_key(hex_key)

    plain_text = None
    if len(hex_str) > 128:
        hex_iv = hex_str[:32]
        hmac_hex_key = hex_str[32:64]
        hmac_hex_str = hex_str[64:128]
        cipher_hex_str = hex_str[128:]

        computed_hmac_hex_str = _compute_HMAC(hex_iv, cipher_hex_str, hex_key, hmac_hex_key)
        if computed_hmac_hex_str.lower() == hmac_hex_str.lower():
            decrypted_str = _decrypt_data(cipher_hex_str, hex_key, hex_iv)
            data = _data_from_hex_string(decrypted_str)
            plain_text = data.decode('utf-8')

    return plain_text


def _compute_HMAC(hex_iv: str, cipher_hex: str, hex_key: str, hmac_hex_key: str):
    hex_key = hex_key.strip().replace(' ', '').lower()
    hmac_hex_key = hmac_hex_key.lower()

    hex_str = hex_iv + cipher_hex + hex_key
    hex_str = hex_str.lower()

    data = hex_str.encode('utf-8')
    hmac_key = hmac_hex_key.encode('utf-8')

    hmac = HMAC.new(hmac_key, digestmod=SHA256)
    hmac.update(data)
    hash_bytes = hmac.digest()

    hash_hex_str = data_to_hex_string(hash_bytes)
    return hash_hex_str


def _check_key(hex_key: str):
    hex_key = hex_key.strip().replace(' ', '').lower()

    if len(hex_key) != 64:
        raise Exception("key length is not 256 bit (64 hex characters)")

    key_len = len(hex_key)
    i = 0
    while i < key_len:
        if hex_key[i] == '0' and hex_key[i+1] == '0':
            raise Exception("key cannot contain zero byte block")

        i += 2


def _encrypt_data(hex_string: str, hex_key: str, hex_iv: str) -> str:
    data = _data_from_hex_string(hex_string)
    key = _data_from_hex_string(hex_key)
    iv = _data_from_hex_string(hex_iv)

    # PyCrypto does not have PKCS7 Padding alforithm, I have to implement on my own.
    #
    # Note: if plain_text length is multiple of 16,
    #        then a sequence of value chr(BS) in length 16 will be appended to the origin data
    data = data + (chr(16 - len(data) % 16) * (16 - len(data) % 16)).encode('utf-8')

    cipher = AES.new(key, AES.MODE_CBC, iv)
    encrypted_data = cipher.encrypt(data)
    encrypted_hex_data = data_to_hex_string(encrypted_data)
    return encrypted_hex_data


def _decrypt_data(hex_string: str, hex_key: str, hex_iv: str) -> str:
    data = _data_from_hex_string(hex_string)
    key = _data_from_hex_string(hex_key)
    iv = _data_from_hex_string(hex_iv)

    cipher = AES.new(key, AES.MODE_CBC, iv)
    decrypted_data = cipher.decrypt(data)

    # PyCrypto does not have PKCS7 Padding alforithm, I have
    #   to implement un-pad on my own.
    decrypted_data = decrypted_data[:-ord(decrypted_data[len(decrypted_data) - 1:])]

    decrypted_hex_data = data_to_hex_string(decrypted_data)
    return decrypted_hex_data

